src/codegen/constant-propagation.lisp

```diff
@@ -0,0 +1,129 @@
+                             (push (cons var propagated-value-node)
+                                   nonconstant-bindings)))
+               (let ((*constant-bindings* (append new-constant-bindings *constant-bindings*)))
+                 (if nonconstant-bindings
+                     (make-node-let
+                      :type (node-type node)
+                      :bindings nonconstant-bindings
+                      :subexpr (funcall *traverse* (node-let-subexpr node) call-stack))
+                     (funcall *traverse* (node-let-subexpr node) call-stack)))))
+
+           (propagate-constants-node-lisp (node call-stack)
+             (declare (type node-lisp node)
+                      (ignore call-stack))
+             (let* ((new-lisp-vars nil)
+                    (let-bindings
+                      (loop :for (lisp-var . coalton-var) :in (node-lisp-vars node)
+                            :for constant-value := (constant-var-value coalton-var :no-error t)
+                            :if constant-value
+                              :collect (let ((new-coalton-var (gentemp (symbol-name coalton-var))))
+                                         (push (cons lisp-var new-coalton-var) new-lisp-vars)
+                                         (cons new-coalton-var constant-value))
+                            :else
+                              :do (push (cons lisp-var coalton-var) new-lisp-vars))))
+               (if let-bindings
+                   (make-node-let
+                    :type (node-type node)
+                    :bindings let-bindings
+                    :subexpr (make-node-lisp
+                              :type (node-type node)
+                              :vars new-lisp-vars
+                              :form (node-lisp-form node)))
+                   (make-node-lisp
+                    :type (node-type node)
+                    :vars new-lisp-vars
+                    :form (node-lisp-form node)))))
+
+           (direct-application-better-infer-types (node call-stack)
+             ;; FIXME: Can we do something similar for non-direct APPLICATION?
+             (let ((constant-propagated-rands
+                     (mapcar (lambda (rand)
+                               (funcall *traverse* rand call-stack))
+                             (node-direct-application-rands node))))
+               (make-node-direct-application
+                :type (node-type node)
+                :rator-type (tc:make-function-type*
+                             (mapcar #'node-type constant-propagated-rands)
+                             (tc:function-return-type (node-type node)))
+                :rator (node-direct-application-rator node)
+                :rands constant-propagated-rands))))
+
+    (let ((*constant-bindings* nil))
+      (traverse
+       node
+       (list
+        (action (:after node-variable) #'propagate-constants-node-variable)
+        (action (:after node-let) #'propagate-constants-node-let)
```

Comment:
For the `node-let` case, you are defining the recursion behavior yourself, calling `*traverse*` on the `node-let-subexpr` with updated `constant-bindings`, so this should be
```suggestion
        (action (:traverse node-let) #'propagate-constants-node-let)
```

(Otherwise, it will traverse once using the default of unchanged args for traversing the bindings and subexpression, and then `:after` that it will do _another_ recursive traversal)